src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,260 @@
+#include "common/protobuf/utility.h"
+#include "envoy/network/connection.h"
+#include "envoy/stats/scope.h"
+#include "src/envoy/tcp/alpn_proxy/alpn_proxy.h"
+#include "src/envoy/tcp/alpn_proxy/alpn_proxy_initial_header.h"
+
+namespace Envoy {
+namespace Tcp {
+namespace AlpnProxy {
+namespace {
+
+std::unique_ptr<::Envoy::Buffer::OwnedImpl> constructProxyHeaderData(
+    const Envoy::ProtobufWkt::Any& proxy_data) {
+  AlpnProxyInitialHeader initial_header;
+  std::string proxy_data_str = proxy_data.SerializeAsString();
+  // Converting from host to network byte order so that most significant byte is
+  // placed first.
+  initial_header.magic = absl::ghtonl(AlpnProxyInitialHeader::magic_number);
+  initial_header.data_size = absl::ghtonl(proxy_data_str.length());
+
+  ::Envoy::Buffer::OwnedImpl initial_header_buffer{
+      absl::string_view(reinterpret_cast<const char*>(&initial_header),
+                        sizeof(AlpnProxyInitialHeader))};
+  auto proxy_data_buffer =
+      std::make_unique<::Envoy::Buffer::OwnedImpl>(proxy_data_str);
+  proxy_data_buffer->prepend(initial_header_buffer);
+  return proxy_data_buffer;
+}
+
+}  // namespace
+
+AlpnProxyConfig::AlpnProxyConfig(const std::string& stat_prefix,
+                                 const std::string& protocol,
+                                 const std::string& node_metadata_id,
+                                 const FilterDirection filter_direction,
+                                 Stats::Scope& scope)
+    : scope_(scope),
+      stat_prefix_(stat_prefix),
+      protocol_(protocol),
+      node_metadata_id_(node_metadata_id),
+      filter_direction_(filter_direction),
+      stats_(generateStats(stat_prefix, scope)) {}
+
+Network::FilterStatus AlpnProxyFilter::onData(Buffer::Instance& data, bool) {
+  switch (conn_state_) {
+    case Invalid:
+    case Done:
+      // No work needed if connection state is Done or Invalid.
+      return Network::FilterStatus::Continue;
+    case ConnProtocolNotRead: {
+      if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+        conn_state_ = Invalid;
+        config_->stats().alpn_protocol_not_found_.inc();
+        return Network::FilterStatus::Continue;
+      } else {
+        conn_state_ = WriteMetadata;
+        config_->stats().alpn_protocol_found_.inc();
+      }
+    }
+    case WriteMetadata: {
+      // TODO(gargnupur): Try to move this just after alpn protocol is
+      // determined and first onData is called in Downstream filter.
+      if (config_->filter_direction_ == FilterDirection::Downstream) {
+        writeNodeMetadata();
+      }
+    }
+    case ReadingInitialHeader:
+    case NeedMoreDataInitialHeader: {
+      tryReadInitialProxyHeader(data);
+      if (conn_state_ == NeedMoreDataInitialHeader) {
+        return Network::FilterStatus::StopIteration;
+      }
+      if (conn_state_ == Invalid) {
+        return Network::FilterStatus::Continue;
+      }
+    }
```

Comment:
Same here. If buffer is large, then we need to drain the proxy header.

Reply:
I thought "return Network::FilterStatus::StopIteration;" is doing that and we don't need to drain "data" and buffer bytes read till now..